api/proto/teleport/machineid/v1/bot_instance.proto

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.machineid.v1;
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "teleport/header/v1/metadata.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/machineid/v1;machineidv1";
+
+// A BotInstance
+message BotInstance {
+  // The kind of resource represented.
+  string kind = 1;
+  // Differentiates variations of the same kind. All resources should
+  // contain one, even if it is never populated.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+  // The configured properties of a BotInstance.
+  BotInstanceSpec spec = 5;
+  // Fields that are set by the server as results of operations. These should
+  // not be modified by users.
+  BotInstanceStatus status = 6;
+}
+
+// BotInstanceSpec contains fields
+message BotInstanceSpec {
+  // Empty as is not user configurable.
```

Comment:
I have some uncertainty with how metadata will be handled on create. The notification service has some prior art here, either [mandating nil metadata](https://github.com/gravitational/teleport/blob/master/lib/services/local/notifications.go#L220) on the incoming resource and extracting the name (uuid) from a field in the spec, or generating the name on demand and [overwriting the `.metadata.name`](https://github.com/gravitational/teleport/blob/master/lib/services/local/notifications.go#L143).

I'd like to avoid exposing the `$botName/$instanceID` string formatting logic to API users, so it might make sense to include `bot_name` as part of `BotInstanceSpec` to allow for this? (and possibly also `id` depending on which component is actually generating the uuid)

Reply:
Hm yeah - tbh - maybe we need to rethink if we even need to Create/Update at the gRPC layer ?

Reply:
Per our offline discussion, I'll be moving `BotName` and `InstanceId` into the `Spec` field and having metadata generated on create. Since it's using the built in generic prefixing functionality that does mean we can't include the bot name in the instance's `.metadata.name`, so it'll just be the UUID. I think that's ultimately for the best though.